File: leetcode/rust/src/21.merge-two-sorted-lists.rs
impl Solution {
    pub fn merge_two_lists(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        //递归 返回当前两个链表的最小的节点
        let some_box_list = |val, next| Some(Box::new(ListNode { val, next }));

        match( l1 , l2){
            (Some(v1), None) => Some(v1), 
            (None, Some(v2)) => Some(v2), 
            (Some(mut v1), Some(mut v2)) => {
                if v1.val < v2.val {
                    some_box_list(v1.val , Self::merge_two_lists(v1.next, Some(v2)))
                }
                else{
                    some_box_list(v2.val , Self::merge_two_lists(v2.next, Some(v1)))
                }
             },
            _ => None
        }

    }
}

File: leetcode/cpp/21.merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution_recursive {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2)
    {
        if (!l1)
            return l2;
        if (!l2)
            return l1;

        if (l1->val > l2->val) {
            l2->next = mergeTwoLists(l1, l2->next);
            return l2;
        } else {
            l1->next = mergeTwoLists(l1->next, l2);
            return l1;
        }
    }
};

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2)
    {
        if (!l1) {
            return l2;
        }
        if (!l2) {
            return l1;
        }

        ListNode* dummy = new ListNode(-1);

        ListNode* current = dummy;

        while (l1 && l2) {
            if (l1->val < l2->val) {
                current->next = l1;
                l1 = l1->next;
                current = current->next;
            } else {
                current->next = l2;
                l2 = l2->next;
                current = current->next;
            }
        }
        if (l1) {
            current->next = l1;
        }

        if (l2) {
            current->next = l2;
        }
        return dummy->next;
    }
};
